packages/shared/src/utils.ts

```diff
@@ -47,17 +52,21 @@ export function getValues(arr: AnyNumArray): NumArray {
   return isNdArray(arr) ? arr.data : arr;
 }
 
-export function toTypedNdArray<
-  T extends NumArray,
-  U extends TypedArrayConstructor
->(arr: NdArray<T>, TypedArrayConstructor: U) {
-  if (isTypedNdArray(arr)) {
+export function toTypedNdArray<T extends TypedArrayConstructor>(
+  arr: NdArray<NumArray>,
+  Constructor: T
+): NdArray<InstanceType<T>> {
+  return ndarray(Constructor.from(arr.data) as InstanceType<T>, arr.shape);
+}
+
+export function toTextureSafeNdArray(
+  arr: NdArray<number[] | TypedArray>
+): NdArray<Float32Array> {
+  if (isTypedNdArray(arr) && isTextureSafeNdArray(arr)) {
     return arr;
```

Comment:
In other words, if `arr` is `NdArray<Float32Array>`, do nothing. I thought the names `toTextureSafeNdArray` and `isTextureSafeNdArray` would better convey the purpose of the functions than `toFloat32NdArray` and `isFloat32NdArray`.

Reply:
For `toTextureSafeNdArray`, I agree but I think renaming `isTextureSafeNdArray` to `isFloat32NdArray` would make this easier to understand